lib/CL/devices/cuda/pocl-ptx-gen.cc

```diff
@@ -0,0 +1,691 @@
+  llvm::GlobalVariable *shared_base = new llvm::GlobalVariable(
+      *module, byte_array_type, false, llvm::GlobalValue::ExternalLinkage, NULL,
+      "_shared_memory_region_", NULL, llvm::GlobalValue::NotThreadLocal, 3,
+      false);
+
+  // Argument info for creating new function
+  std::vector<llvm::Argument *> arguments;
+  std::vector<llvm::Type *> argument_types;
+
+  // Loop over arguments
+  bool has_local_args = false;
+  for (auto arg = function->arg_begin(); arg != function->arg_end(); arg++) {
+    // Check for local memory pointer
+    llvm::Type *arg_type = arg->getType();
+    if (arg_type->isPointerTy() && arg_type->getPointerAddressSpace() == 3) {
+      has_local_args = true;
+
+      // Create new argument for offset into shared memory allocation
+      llvm::Type *i32ty = llvm::Type::getInt32Ty(context);
+      llvm::Argument *offset =
+          new llvm::Argument(i32ty, arg->getName() + "_offset");
+      arguments.push_back(offset);
+      argument_types.push_back(i32ty);
+
+      // Insert GEP to add offset
+      llvm::Value *zero = llvm::ConstantInt::getSigned(i32ty, 0);
+      llvm::GetElementPtrInst *gep = llvm::GetElementPtrInst::Create(
+          byte_array_type, shared_base, {zero, offset});
+      gep->insertBefore(&*function->begin()->begin());
+
+      // Cast pointer to correct type
+      llvm::BitCastInst *cast = new llvm::BitCastInst(gep, arg_type);
+      cast->insertAfter(gep);
+
+      cast->takeName(&*arg);
+      arg->replaceAllUsesWith(cast);
+    } else {
+      // No change to other arguments
+      arguments.push_back(&*arg);
+      argument_types.push_back(arg_type);
+    }
+  }
+
+  if (!has_local_args)
+    return;
+
+  // Create new function with offsets instead of local memory pointers
+  llvm::FunctionType *new_func_type =
+      llvm::FunctionType::get(function->getReturnType(), argument_types, false);
+  llvm::Function *new_func = llvm::Function::Create(
+      new_func_type, function->getLinkage(), function->getName(), module);
+  new_func->takeName(function);
+
+  // Take function body from old function
+  new_func->getBasicBlockList().splice(new_func->begin(),
+                                       function->getBasicBlockList());
+
+  // TODO: Copy attributes from old function
+
+  // Update function body with new arguments
+  std::vector<llvm::Argument *>::iterator old_arg;
+  llvm::Function::arg_iterator new_arg;
+  for (old_arg = arguments.begin(), new_arg = new_func->arg_begin();
+       new_arg != new_func->arg_end(); new_arg++, old_arg++) {
+    new_arg->takeName(*old_arg);
+    (*old_arg)->replaceAllUsesWith(&*new_arg);
+  }
+
+  // TODO: Deal with calls to this kernel from other function?
+
+  function->eraseFromParent();
+}
+
+void pocl_insert_ptx_intrinsics(llvm::Module *module) {
+  struct ptx_intrinsic_map_entry {
+    const char *varname;
+    const char *intrinsic;
+  };
+  struct ptx_intrinsic_map_entry intrinsic_map[] = {
```

Comment:
We just discussed an issue related to this with Michal. We are doing this backwards in pocl (legacy code). We should use builtins to start with which can be then lowered to whatever the target wants (by default to use the hidden context struct).